drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,522 @@
+/*
+ * Copyright (c) 2024 Charles Dias <charlesdias.cd@outlook.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_dcmi
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/video.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/dma/dma_stm32.h>
+
+#include <stm32_ll_dma.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(video_stm32_dcmi, CONFIG_STM32_DCMI_LOG_LEVEL);
+
+#ifndef CONFIG_SOC_SERIES_STM32H7X
+	#error "This driver version is only supported on STM32H7X series"
+#endif
+
+K_HEAP_DEFINE(video_stm32_buffer_pool,
+		CONFIG_VIDEO_BUFFER_POOL_SZ_MAX);
+
+typedef void (*irq_config_func_t)(const struct device *dev);
+
+struct stream {
+	DMA_TypeDef *reg;
+	const struct device *dma_dev;
+	uint32_t channel;
+	struct dma_config cfg;
+};
+
+struct video_stm32_dcmi_data {
+	const struct device *dev;
+	DCMI_HandleTypeDef hdcmi;
+	struct stream dma;
+	struct video_format fmt;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	uint32_t pixel_format;
+	uint32_t height;
+	uint32_t width;
+	uint32_t pitch;
+	uint8_t *buffer;
+};
+
+struct video_stm32_dcmi_config {
+	struct stm32_pclken pclken;
+	irq_config_func_t irq_config;
+	const struct pinctrl_dev_config *pctrl;
+	const struct device *sensor_dev;
+};
+
+static inline unsigned int video_pix_fmt_bpp(uint32_t pixelformat)
+{
+	switch (pixelformat) {
+	case VIDEO_PIX_FMT_BGGR8:
+	case VIDEO_PIX_FMT_GBRG8:
+	case VIDEO_PIX_FMT_GRBG8:
+	case VIDEO_PIX_FMT_RGGB8:
+		return 1;
+	case VIDEO_PIX_FMT_RGB565:
+	case VIDEO_PIX_FMT_YUYV:
+		return 2;
+	default:
+		return 0;
+	}
+}
+
+void HAL_DCMI_ErrorCallback(DCMI_HandleTypeDef *hdcmi)
+{
+	LOG_WRN("%s", __func__);
+}
+
+void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
+{
+	HAL_DCMI_Suspend(hdcmi);
```

Comment:
Could be moved after local variables definitions ?

Reply:
Done.